### PR TITLE
Add lint to detect `allow` attributes without reason

### PR DESCRIPTION
**Pull request by xFrednet**
I was considering putting this lint into the pedantic group. However, that would result in countless warnings for existing projects. Having it in restriction also seems good to me :upside_down_face: (And now I need sleep :zzz: )

---

changelog: New lint [`allow_lint_without_reason`] (Requires the `lint_reasons` feature)

Closes: rust-lang/rust-clippy#8502

### CODE REVIEW

File: clippy_lints/src/attrs.rs [hunk 1]
```diff
@@ -255,7 +255,38 @@ declare_clippy_lint! {
...
+    /// ```
+    ///
+    /// Good:
+    /// ```rust
+    /// #![feature(lint_reasons)]
+    ///
+    /// #![allow(clippy::some_lint, reason = "False positive rust-lang/rust-clippy#1002020")]
+    /// ```
+    #[clippy::version = "1.61.0"]
+    pub ALLOW_LINT_WITHOUT_REASON,
```
Comment (flip1995) [inline clippy_lints/src/attrs.rs:283]: Lint name suggestion:
```suggestion
    pub ALLOW_ATTRIBUTES_WITHOUT_REASON,
```
This flows better with `allow(clippy::allow_attributes_without_reason)`. Still not perfect, but better than with the current lint name IMO

File: clippy_lints/src/attrs.rs [hunk 2]
```diff
@@ -404,6 +438,30 @@ fn check_clippy_lint_names(cx: &LateContext<'_>, name: Symbol, items: &[NestedMe
...
+        && item.path == sym::reason
+    {
+        return;
+    }
+
+    span_lint_and_help(
+        cx,
+        ALLOW_LINT_WITHOUT_REASON,
+        attr.span,
+        &format!("`{}` attribute without reason", name.as_str()),
```
Comment (flip1995) [inline clippy_lints/src/attrs.rs:459]:
```suggestion
        &format!("`{}` attribute without specifying a reason", name.as_str()),
```

maybe?

**Review by flip1995 — COMMENTED** [review]
LGTM overall. `restriction` sounds like the orrect lint group to me :+1:

**Comment by xFrednet** [comment]
Good suggestions :+1: The lint name was also something I was thinking about :upside_down_face: I've added your suggestions and squashed the commits :upside_down_face:

**Comment by flip1995** [comment]
@bors r+

Thanks!

**Comment by matthiaskrgr** [comment]
@bors retry
````

Run actions/download-artifact@v1
Downloading artifact 'target' to: '/home/runner/work/rust-clippy/rust-clippy/target'
Error: The HTTP request timed out after 00:01:40.
Error: Exit code 1 returned from process: file name '/home/runner/runners/2.288.1/bin/Runner.PluginHost', arguments 'action "GitHub.Runner.Plugins.Artifact.DownloadArtifact, Runner.Plugins"'.
````

**Pull request merged**